mitodl/mit-open
# Address performance of learning resource APIs

### What are the relevant tickets?




Part of https://github.com/mitodl/hq/issues/4830

### Description (What does it do?)


This PR formalizes how we prefetch learning resources by adding custom `QuerySet` implementations with a standard `for_serialization` method. Going forward this will be where all prefetches that are necessary to query for and serialize model objects efficiently. We should not be performing database queries in serializers going forward (aside from validation, `create()`, and `update()`).

A lot of code got moved out of serializers and into `@cached_property` on django models since this is far more robust and aligns to django's practices of how to write models.

An added benefit that I saw from this was by applying the `for_serialization` to the indexing operations I saw it go from 2k queries down to ~70. `ContentFile` indexing similarly reduced from 100-200 to a few dozen. It's not really possible to assert query counts on these because there's some variation in query counts I couldn't figure out and it's out of scope anyway.

### How can this be tested?

- The APIs should return the same data as on `main` but be much faster.
- The search indexing should similarly serialize the data the same but will probably run much faster as well.
